File: solutions/python/arrays-hashing/word-pattern.py
# LeetCode 290 — Word Pattern (Easy)
# Category: Arrays & Hashing · Approach: Two Maps
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/word-pattern/

def word_pattern(pattern: str, s: str) -> bool:
    words = s.split()
    if len(pattern) != len(words):
        return False
    p_to_w, w_to_p = {}, {}
    for p, w in zip(pattern, words):
        if p in p_to_w or w in w_to_p:
            if p_to_w.get(p) != w or w_to_p.get(w) != p:
                return False
        else:
            p_to_w[p] = w
            w_to_p[w] = p
    return True

File: solutions/typescript/arrays-hashing/word-pattern.ts
// LeetCode 290 — Word Pattern (Easy)
// Category: Arrays & Hashing · Approach: Two Maps
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/word-pattern/

function wordPattern(pattern: string, s: string): boolean {
  const words = s.split(" ");
  if (pattern.length !== words.length) return false;
  const pToW = new Map<string, string>();
  const wToP = new Map<string, string>();
  for (let i = 0; i < pattern.length; i++) {
    const p = pattern[i], w = words[i];
    if (pToW.has(p) || wToP.has(w)) {
      if (pToW.get(p) !== w || wToP.get(w) !== p) return false;
    } else {
      pToW.set(p, w);
      wToP.set(w, p);
    }
  }
  return true;
}
